EDDA Normalization 
========================================================
## Author: Sumaiya Islam
## Date: November 12, 2015


### A. Set up wd & packages

Use R version 3.1.1 (2014-07-10)

Here we will continue from pre-processing of our samples to the normalization of our samples using the objects EDDA.RData and EDDA.2.RData.

Initially we will set our working directory and load our libraries:
```{r EDDA_setup, include = FALSE, echo=FALSE}
setwd("/home/sislam/EDDA")
library(methylumi)
library(gplots)
library(marray)
library(lumi)
library(lattice)
library("RColorBrewer")
library(knitr)
library(xtable)
library(wateRmelon)
library(limma)
library(sva)
```

### B. Load files

Only two files need to be loaded before we can begin:
- **EDDA**, this is the MethylLumiM object which has been pre-processed and should contain 452,266 probes (in our particular case) and 72 samples.
- **EDDA.2**, is the MethyLumiSet object which is needed for the normalization process and has not yet been utilized.

```{r EDDA_fileLoad, echo=FALSE}
load("EDDA(filtered).RData")
load("EDDA.2.RData")
dim(EDDA)
dim(EDDA.2)
```
**EDDA** should contain 452,266 probes and 72 samples where **EDDA.2** should contain 485,577 probes and 72 samples.


### C. Quantro

The quantro R-package (Hicks S and Irizarry R (2015). quantro: A test for when to use quantile normalization. *Genome Biol*) can be used to test for global differences between groups of distributions which assess whether global normalization methods such as quantile normalization should be applied.

```{r}
library(quantro)
EDDA_betas<-as.matrix(betas(EDDA))
matboxplot(EDDA_betas, groupFactor = EDDA$Sample_Group, col = c(2,3), xaxt = "n",main = "Beta Values")
```



```{r}
library(doParallel)
registerDoParallel(cores=4)
qtestPerm <- quantro(EDDA_betas, EDDA$Sample_Group, B = 1000)
qtestPerm
quantroPlot(qtestPerm)

```

We can see that the quantro stat (1.17071) is significant (p > 0.05), indicating that there are global differences in distributions across groups so quantile normalization is appropriate for this dataset.



### D. Quantile normalization

 The **lumiMethylN** is the lumi package's quantile normalization function.

```{r EDDA_luminormalization, include=FALSE, echo=FALSE}
# all samples
EDDA.q <- lumiMethyN(EDDA, method = "quantile")
```

### E. Normalization: SWAN

The 450K Illumina Infinium Array has inherent variation associated with its methodologies which must be accounted for in our analyses. Much of this variation is attributed to the use of two types of probes used in the array, Type I and Type II.

Type I probes contain two bead types corresponding to an unmethylated (U) or methylated (M) status. Type I probes obtain methylation status as a result of fluoresence expressed after a single base pair extension occurs just after the target basepair, resulting in only one color channel being utilized (red). Type I probes also assume that any CpG sites underlying the probe are of the same status as the query site (methylated or unmethylated). The beta values for Type I probes are then determined by this formula b= M/(U + M). Type II probes do not assume the methylation status of underlying CpG sites within the probe and so consist of a combination of degenerate probes with varying combinations of up to three underlying CpG sites. Type II probes also detect methylation status with a single base pair extension except that the site being extended is the CpG site of detection and so require two fluorescent colors green for methylated (M) and red for unmethylated (U) sites. Type II probe beta values are calculated using this formula b = Green (M)/(Red (U) + Green (M)). 

Due to these inherent differences between Type I and Type II probes used in the Illumina Infinium 450K array several groups in the field have deveolped various normalization analyses to correct for the differences between these probes. 

#### Subset-quantile within array normalization (SWAN)

SWAN normalization sets out to correct for differences associated by the use of the two different probe types. As the beta value distributions vary between the two probe types, resulting from the use of different color channels, SWAN attempts to normalize the data by reducing these differences. This is accomplished by performing quantile normalization on subsets of probes stratified based on the number of underlying CpGs present within the probes. The reason for this stratification is SWAN takes into account that there are inherent biological differences between regions which are CpG dense and CpG sparse and so tries to accomplish normalizing technical variation without interferring with true biological variation (Maksimovic *et. al* 2012 *Genome Biology*). 


```{r EDDA_SWAN, echo=FALSE}
runSWAN<- function(x,y){
  target <- fData(QCdata(x))$TargetID =="NEGATIVE"
  grn <- methylated(QCdata(x))[target,]
  red <- unmethylated(QCdata(x))[target,]
  bkg <- rowMeans(cbind(colMeans(grn),colMeans(red)))
  all(names(bkg)%in%sampleNames(y)) 
  bkg <- bkg[sampleNames(y)]
  source("SWAN scripts.R")
  load("nCpG all.RData")
  all(featureNames(y)%in%names(nCpG.all))
  fData(y)$nCpG <- nCpG.all[featureNames(y)]
  methylumi.swan <- preprocessMSWAN(y,bg=bkg)
  return(methylumi.swan)
  }

# all samples
EDDA.swan<-runSWAN(EDDA.2,EDDA.q)
# checks
sum(is.na(exprs(EDDA.swan))) # should be 0
sum(is.na(unmethylated(EDDA.swan))) # should be 0
sum(is.na(methylated(EDDA.swan))) # should be 0
```

### Comparing raw, quantile vs SWAN-normalized density plots

For all samples
```{r EDDA_compareAll, echo=FALSE}

groupCol <- as.numeric(factor(EDDA$Sample_Group))
groupCol <- gsub("1", "orange",  gsub("2", "blue", gsub("3", "green", groupCol)))
par(mfrow = c(1, 3))
density(EDDA,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Raw", col=groupCol)
density(EDDA.q,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Quantile Normalized", col=groupCol)
density(EDDA.swan,legend=NULL, lty = 1, xlim = c(-6, 6), main = "SWAN Normalized", col=groupCol)
```


We can see that there is a dramatic reduction in sample-to-sample variability from the raw data to SWAN-normalized in all the datasets. 

 
### F. Plot beta values before/after normalization

We will show the distribution of beta values for the entire dataset (all samples normalized together) in its raw form and final normalized form.

```{r EDDA_betaPlot, echo=FALSE} 
type <- fData(EDDA.q)$INFINIUM_DESIGN_TYPE

before.betas <- betas(EDDA)
after.betas.SWAN <- betas(EDDA.swan)

par(mfrow = c(1, 1))
plot(c(0, 1) , c(0, 10), type = "n", xlab = "Beta values", ylab = "Density")
lines(density(na.omit(before.betas)[type=="I",]),col="blue",lty=1)
lines(density(na.omit(before.betas)[type=="II",]),col="blue",lty=2, wld=1.5)
lines(density(na.omit(after.betas.SWAN)[type=="I",]),col="red",lty=1)
lines(density(na.omit(after.betas.SWAN)[type=="II",]),col="red",lty=2, lwd=1.5)
legend("top", inset=.05, cex=0.6, c("Raw Type 1","Raw Type 2", "Normalized Type 1", "Normalized Type 2"), col=c("blue","blue","red", "red"), lty=c(1,2,1,2), horiz=FALSE)
```

We can see that there is improved peak-to-peak overlap of the Type 1 and Type 2 probes with the final normalized dataset as desired (ie Type 1 and 2 probes have more similar dynamic ranges for their beta values). 

### G. Save normalized data

```{r EDDA_normalized_save, echo=FALSE}
# save(EDDA.swan, file = "EDDA.swan.RData")
```


### H. Check Technical Replicate Correlation

We determined from the 65 SNP profile that EDDA_59 and EDDA_70 are likely technical replicates (same DNA submitted twice). We will use these as technical replicates to check for technical replicate correlations pre and post-normalization. We expect improved correlation between replicates after normalization. 


```{r}
RepCor  <- function(x,y) {
  sRep <- x[, sampleNames(x)%in%y]
  cor.dat<- cor(betas(sRep), use = "pairwise.complete.obs")
  return(result = cor.dat[1, 2])
}

Replicates<-c("EDDA_59", "EDDA_70")

(pre_norm_correlation<-RepCor(EDDA, Replicates)) # 0.9979
(post_norm_correlation<-RepCor(EDDA.swan, Replicates)) # 0.9982

```

We can see there is a slight improvement in correlation between technical replicates (from 0.9979 to 0.9982) as desired. 

We will continue on with our analysis using the SWAN-normalized dataset.